{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- GRPH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rosalind_9696 Rosalind_0849\n",
      "Rosalind_0905 Rosalind_8584\n",
      "Rosalind_9578 Rosalind_3045\n",
      "Rosalind_9578 Rosalind_2947\n",
      "Rosalind_1665 Rosalind_9475\n",
      "Rosalind_1665 Rosalind_4396\n",
      "Rosalind_1665 Rosalind_6445\n",
      "Rosalind_1665 Rosalind_9127\n",
      "Rosalind_1045 Rosalind_5144\n",
      "Rosalind_0471 Rosalind_3412\n",
      "Rosalind_0471 Rosalind_9296\n",
      "Rosalind_3412 Rosalind_6115\n",
      "Rosalind_3412 Rosalind_4019\n",
      "Rosalind_3412 Rosalind_5612\n",
      "Rosalind_7625 Rosalind_4983\n",
      "Rosalind_7625 Rosalind_7544\n",
      "Rosalind_7625 Rosalind_4256\n",
      "Rosalind_2725 Rosalind_3412\n",
      "Rosalind_2725 Rosalind_9296\n",
      "Rosalind_6115 Rosalind_4700\n",
      "Rosalind_6115 Rosalind_0028\n",
      "Rosalind_8201 Rosalind_8831\n",
      "Rosalind_8124 Rosalind_9868\n",
      "Rosalind_8124 Rosalind_8649\n",
      "Rosalind_9475 Rosalind_9927\n",
      "Rosalind_7066 Rosalind_1045\n",
      "Rosalind_7066 Rosalind_7954\n",
      "Rosalind_7066 Rosalind_9232\n",
      "Rosalind_7066 Rosalind_6794\n",
      "Rosalind_3089 Rosalind_3265\n",
      "Rosalind_8148 Rosalind_8859\n",
      "Rosalind_8148 Rosalind_4961\n",
      "Rosalind_2392 Rosalind_4700\n",
      "Rosalind_2392 Rosalind_0028\n",
      "Rosalind_4648 Rosalind_6912\n",
      "Rosalind_3045 Rosalind_8236\n",
      "Rosalind_2789 Rosalind_9927\n",
      "Rosalind_9296 Rosalind_3045\n",
      "Rosalind_9296 Rosalind_2947\n",
      "Rosalind_2947 Rosalind_1045\n",
      "Rosalind_2947 Rosalind_7954\n",
      "Rosalind_2947 Rosalind_9232\n",
      "Rosalind_2947 Rosalind_6794\n",
      "Rosalind_7185 Rosalind_9696\n",
      "Rosalind_7185 Rosalind_0905\n",
      "Rosalind_8859 Rosalind_4532\n",
      "Rosalind_0368 Rosalind_3265\n",
      "Rosalind_6445 Rosalind_7625\n",
      "Rosalind_6445 Rosalind_0853\n",
      "Rosalind_6445 Rosalind_8151\n",
      "Rosalind_6445 Rosalind_8916\n",
      "Rosalind_5239 Rosalind_8889\n",
      "Rosalind_9303 Rosalind_3840\n",
      "Rosalind_9303 Rosalind_7794\n",
      "Rosalind_9303 Rosalind_3666\n",
      "Rosalind_8917 Rosalind_6469\n",
      "Rosalind_8917 Rosalind_5472\n",
      "Rosalind_0646 Rosalind_7184\n",
      "Rosalind_0849 Rosalind_4700\n",
      "Rosalind_0849 Rosalind_0028\n",
      "Rosalind_9127 Rosalind_9303\n",
      "Rosalind_9127 Rosalind_8766\n",
      "Rosalind_3834 Rosalind_9578\n",
      "Rosalind_7901 Rosalind_6912\n",
      "Rosalind_7436 Rosalind_8584\n",
      "Rosalind_3760 Rosalind_0471\n",
      "Rosalind_3548 Rosalind_0471\n",
      "Rosalind_4578 Rosalind_8917\n",
      "Rosalind_4578 Rosalind_6417\n",
      "Rosalind_4578 Rosalind_3205\n",
      "Rosalind_1252 Rosalind_9927\n",
      "Rosalind_9868 Rosalind_7185\n",
      "Rosalind_9868 Rosalind_2592\n",
      "Rosalind_8629 Rosalind_8164\n",
      "Rosalind_8629 Rosalind_3992\n",
      "Rosalind_9684 Rosalind_6912\n",
      "Rosalind_4983 Rosalind_2789\n",
      "Rosalind_4983 Rosalind_8338\n",
      "Rosalind_4983 Rosalind_3072\n",
      "Rosalind_7794 Rosalind_3412\n",
      "Rosalind_7794 Rosalind_9296\n",
      "Rosalind_5144 Rosalind_0849\n",
      "Rosalind_7544 Rosalind_3760\n",
      "Rosalind_7954 Rosalind_9696\n",
      "Rosalind_7954 Rosalind_0905\n",
      "Rosalind_2951 Rosalind_3045\n",
      "Rosalind_2951 Rosalind_2947\n",
      "Rosalind_8164 Rosalind_9303\n",
      "Rosalind_8164 Rosalind_8766\n",
      "Rosalind_3992 Rosalind_8236\n",
      "Rosalind_8338 Rosalind_3265\n",
      "Rosalind_0853 Rosalind_4532\n",
      "Rosalind_7255 Rosalind_0842\n",
      "Rosalind_7255 Rosalind_7985\n",
      "Rosalind_4961 Rosalind_6777\n",
      "Rosalind_8831 Rosalind_9868\n",
      "Rosalind_8831 Rosalind_8649\n",
      "Rosalind_5463 Rosalind_1045\n",
      "Rosalind_5463 Rosalind_7954\n",
      "Rosalind_5463 Rosalind_9232\n",
      "Rosalind_5463 Rosalind_6794\n",
      "Rosalind_8766 Rosalind_6777\n",
      "Rosalind_0860 Rosalind_8584\n",
      "Rosalind_4019 Rosalind_2725\n",
      "Rosalind_4019 Rosalind_3548\n",
      "Rosalind_4804 Rosalind_8831\n",
      "Rosalind_5533 Rosalind_9868\n",
      "Rosalind_5533 Rosalind_8649\n",
      "Rosalind_5612 Rosalind_2725\n",
      "Rosalind_5612 Rosalind_3548\n",
      "Rosalind_6469 Rosalind_3834\n",
      "Rosalind_6469 Rosalind_9334\n",
      "Rosalind_0842 Rosalind_8859\n",
      "Rosalind_0842 Rosalind_4961\n",
      "Rosalind_0028 Rosalind_9696\n",
      "Rosalind_0028 Rosalind_0905\n",
      "Rosalind_8151 Rosalind_7901\n",
      "Rosalind_8236 Rosalind_4983\n",
      "Rosalind_8236 Rosalind_7544\n",
      "Rosalind_8236 Rosalind_4256\n",
      "Rosalind_9232 Rosalind_9051\n",
      "Rosalind_9440 Rosalind_0842\n",
      "Rosalind_9440 Rosalind_7985\n",
      "Rosalind_6777 Rosalind_9696\n",
      "Rosalind_6777 Rosalind_0905\n",
      "Rosalind_9624 Rosalind_7184\n",
      "Rosalind_3265 Rosalind_6115\n",
      "Rosalind_3265 Rosalind_4019\n",
      "Rosalind_3265 Rosalind_5612\n",
      "Rosalind_5472 Rosalind_6115\n",
      "Rosalind_5472 Rosalind_4019\n",
      "Rosalind_5472 Rosalind_5612\n",
      "Rosalind_8916 Rosalind_9696\n",
      "Rosalind_8916 Rosalind_0905\n",
      "Rosalind_7362 Rosalind_6115\n",
      "Rosalind_7362 Rosalind_4019\n",
      "Rosalind_7362 Rosalind_5612\n",
      "Rosalind_7985 Rosalind_1045\n",
      "Rosalind_7985 Rosalind_7954\n",
      "Rosalind_7985 Rosalind_9232\n",
      "Rosalind_7985 Rosalind_6794\n",
      "Rosalind_8584 Rosalind_6469\n",
      "Rosalind_8584 Rosalind_5472\n",
      "Rosalind_4532 Rosalind_5144\n",
      "Rosalind_9927 Rosalind_3265\n",
      "Rosalind_2592 Rosalind_4532\n",
      "Rosalind_6082 Rosalind_0368\n",
      "Rosalind_6082 Rosalind_6456\n",
      "Rosalind_6082 Rosalind_0860\n",
      "Rosalind_6794 Rosalind_0842\n",
      "Rosalind_6794 Rosalind_7985\n",
      "Rosalind_4256 Rosalind_6115\n",
      "Rosalind_4256 Rosalind_4019\n",
      "Rosalind_4256 Rosalind_5612\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "input_file = 'rosalind_grph.txt'\n",
    "k=3\n",
    "with open(input_file) as f:\n",
    "    fasta_sequences = list(SeqIO.parse(f, 'fasta'))\n",
    "\n",
    "for fasta1 in fasta_sequences:\n",
    "    for fasta2 in fasta_sequences:\n",
    "        n1=fasta1.id\n",
    "        first =str(fasta1.seq)\n",
    "        n2= fasta2.id\n",
    "        second = str(fasta2.seq)\n",
    "        if first == second:\n",
    "            continue\n",
    "        suffix = first[-k:]\n",
    "        prefix = second[:k]\n",
    "        if prefix !=suffix:\n",
    "            continue\n",
    "        else:\n",
    "            print(n1, n2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- orf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MVISHSSRQEGFRFCENRKARGWIVDILDVMRARAKSYR\n",
      "MDRRHTRRYACASKIVSLASPAASFNR\n",
      "MRARAKSYR\n",
      "MRQT\n",
      "MCLIRSARGALKCARGVQKSSTDNDIRWATSMSSTA\n",
      "MCPRRAEVFY\n",
      "MISGGLLV\n",
      "MSSTA\n",
      "MALPVDFRLIRASLKRLYVLKQNT\n",
      "MS\n",
      "MKTLKS\n",
      "MRRLLRFSPDEPHRILIIDYSTCRSACRR\n",
      "MNPTGSLLSIIAPVVQLVAGDYTVYLLKPCGDRISLDYIWLDYILEVRLVQTRSREAIGCSSNH\n",
      "MAGLYS\n",
      "MLE\n",
      "MYGPGAIL\n",
      "MPFERMDGDRLLVYRPMSWQRAVRVPSDLHCG\n",
      "MDGDRLLVYRPMSWQRAVRVPSDLHCG\n",
      "MSWQRAVRVPSDLHCG\n",
      "MNHPQCKSLGTLTARCQDIGR\n",
      "MEASQYGTWSVH\n",
      "MAPGPYIRRVNGYSSIQLPHASGSVPDEPQEYSPAIYNPS\n",
      "MVTRASNCLTRAGLYQTNLKNIVQPYIIQADPIPTGL\n",
      "MVLGVLFQDIQPL\n",
      "MRSFLLKGDLR\n",
      "MVFFGRLV\n",
      "MNGTSSR\n",
      "MAPALGDQAVLLILVAHLISLSVEDFCTPRAHFSAPRADRIKHI\n",
      "MLTIRFCSRTHNV\n",
      "M\n"
     ]
    }
   ],
   "source": [
    "def rev_comp(fa):\n",
    "    \n",
    "    fa_transcript = ''\n",
    "    for i in range(len(fa)):\n",
    "        index = -(i+1)\n",
    "        if fa[index] == 'A':\n",
    "            fa_transcript = fa_transcript + 'U'\n",
    "        elif fa[index] == 'T':\n",
    "            fa_transcript = fa_transcript + 'A'\n",
    "        elif fa[index] == 'C':\n",
    "            fa_transcript = fa_transcript + 'G'\n",
    "        else:\n",
    "            fa_transcript = fa_transcript +'C'\n",
    "    fa_rc = ''\n",
    "    for i in range(len(fa_transcript)):\n",
    "        index = -(i+1)\n",
    "        if fa_transcript[index] == 'A':\n",
    "            fa_rc = fa_rc + 'U'\n",
    "        elif fa_transcript[index] == 'U':\n",
    "            fa_rc = fa_rc + 'A'\n",
    "        elif fa_transcript[index] == 'C':\n",
    "            fa_rc = fa_rc + 'G'\n",
    "        else:\n",
    "            fa_rc = fa_rc +'C'\n",
    "    return fa_transcript, fa_rc\n",
    "def find_all(s,substring):\n",
    "   \n",
    "    index_list = []\n",
    "    index = s.find(substring)\n",
    "    while index != -1: \n",
    "        index_list.append(index)\n",
    "        index = s.find(substring, index+1)\n",
    "    \n",
    "    if len(index_list) > 0:\n",
    "        return index_list\n",
    "    else:\n",
    "        return -1\n",
    "\n",
    "def orf(mrna):\n",
    "    \n",
    "    start_codon = 'AUG'\n",
    "    stop_codon = ['UAA', 'UAG', 'UGA']\n",
    "    i, j = 0,0\n",
    "    out = []\n",
    "    while i <= len(mrna)-2:\n",
    "        if mrna[i:i+3] == start_codon:\n",
    "            j=i\n",
    "            sequence=''\n",
    "            while i<= len(mrna) -2:\n",
    "                if mrna[i:i+3] in stop_codon:\n",
    "                    out.append(sequence)\n",
    "                    break\n",
    "                sequence = sequence + mrna[i:i+3]\n",
    "                i = i+3\n",
    "        i = j+1\n",
    "        j = j+1\n",
    "    \n",
    "    return out\n",
    "\n",
    "def translate(rnaseq):\n",
    "    codon_table = { 'UUU': 'F', 'CUU': 'L', 'AUU': 'I', 'GUU': 'V', \\\n",
    "                    'UUC': 'F', 'CUC': 'L', 'AUC': 'I', 'GUC': 'V', \\\n",
    "                    'UUA': 'L', 'CUA': 'L', 'AUA': 'I', 'GUA': 'V', \\\n",
    "                    'UUG': 'L', 'CUG': 'L', 'AUG': 'M', 'GUG': 'V', \\\n",
    "                    'UCU': 'S', 'CCU': 'P', 'ACU': 'T', 'GCU': 'A', \\\n",
    "                    'UCC': 'S', 'CCC': 'P', 'ACC': 'T', 'GCC': 'A', \\\n",
    "                    'UCA': 'S', 'CCA': 'P', 'ACA': 'T', 'GCA': 'A', \\\n",
    "                    'UCG': 'S', 'CCG': 'P', 'ACG': 'T', 'GCG': 'A', \\\n",
    "                    'UAU': 'Y', 'CAU': 'H', 'AAU': 'N', 'GAU': 'D', \\\n",
    "                    'UAC': 'Y', 'CAC': 'H', 'AAC': 'N', 'GAC': 'D', \\\n",
    "                    'UAA': 'Stop', 'CAA': 'Q', 'AAA': 'K', 'GAA': 'E', \\\n",
    "                    'UAG': 'Stop', 'CAG': 'Q', 'AAG': 'K', 'GAG': 'E', \\\n",
    "                    'UGU': 'C', 'CGU': 'R', 'AGU': 'S', 'GGU': 'G', \\\n",
    "                    'UGC': 'C', 'CGC': 'R', 'AGC': 'S', 'GGC': 'G', \\\n",
    "                    'UGA': 'Stop', 'CGA': 'R', 'AGA': 'R', 'GGA': 'G', \\\n",
    "                    'UGG': 'W', 'CGG': 'R', 'AGG': 'R', 'GGG': 'G'}\n",
    "    length = len(rnaseq)\n",
    "    proseq = []\n",
    "    for i in range(0,length,3):\n",
    "        triplet = rnaseq[i:i+3]\n",
    "        if codon_table[str(triplet)] != 'Stop':\n",
    "            proseq.append(codon_table[str(triplet)])\n",
    "        else:\n",
    "            break\n",
    "    proseq = ''.join(proseq)\n",
    "    return proseq\n",
    "\n",
    "\n",
    "\n",
    "dna = 'GCAGCGACCAACTATAATCTGATTCCAACTAGTGAGAATTCTTCTGGTCCCTCGGTCTAATCTGGGTTGATTGAGTTTTCCTGCCAACCTGACGAACAGATGAATCACCCACAGTGTAAATCACTGGGCACCCTGACGGCTCGTTGCCAAGACATCGGACGATAGACAAGGAGTCGATCTCCGTCCATACGCTCAAAGGGCATGGAGGCGTCACAGTATGGCACCTGGTCCGTACATTAGGCGTGTCAATGGTTACTCGAGCATCCAATTGCCTCACGCGAGCGGGTCTGTACCAGACGAACCTCAAGAATATAGTCCAGCCATATATAATCCAAGCTGATCCGATCCCCACAGGGCTTTAGTAGGTATACCGTGTAGTCACCGGCGACAAGCTGAACGACAGGTGCTATAATCGATAATAAGGATCCTGTGGGGTTCATCGGGACTAAATCTAAGTAACCTTCGCATTAGGAGGATGAAGACTTTGAAATCCTAGCTAGGATTTCAAAGTCTTCATGGTTCTAGGTGTTTTGTTTCAAGACATACAGCCGCTTTAAACTCGCCCTGATTAATCGAAAGTCCACGGGCAAAGCCATGCGTAGCTTTTTACTGAAGGGTGACTTGAGGTAACGAATTCTTGTATGGTTTTTTTTGGAAGGCTAGTTTAGGTATTAGCTGTGCAACTAGCCGATGAATGGCACCAGCTCTCGGTGATCAGGCCGTGCTGCTCATACTAGTAGCCCACCTGATATCATTGTCAGTAGAAGACTTCTGCACGCCTCGGGCACATTTTAGTGCTCCCCGGGCAGACCGAATTAAACACATTTAAGTTTGACGCATTACCGGTTGAAGCTAGCAGCAGGAGATGCTAACGATACGATTTTGCTCGCGCACGCATAACGTCTAGTATGTCTACGATCCATCCTCGTGCTTTGCGATTTTCGCAAAAGCGGAAGCCCTCCTGCCGACTGCTATGTGAGATTACCATGAGA'\n",
    "mrna1, mrna2 = rev_comp(dna)\n",
    "oprf1=orf(mrna1)\n",
    "oprf2=orf(mrna2)\n",
    "orf_list = oprf1\n",
    "for i in oprf2:\n",
    "    if i not in orf_list:\n",
    "        orf_list.append(i)\n",
    "\n",
    "for i in orf_list:\n",
    "    print(translate(i))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- RNA SPLICING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MRQGKRYVCRNCDRVIFPMAWAFSQETPIALATRKSTCRYRRIVNSGGGPAFELHRPNVLQCPVHVCSCRSRLPHQNTDNARSSLLKCARRASGQSAVRIRSVVPIYRKTTFTTDEPGDVNYTRTLHRYLGLNTTTRSLHQIQHQLRPIDRPPRNATIKFGAGRISYCNVEACTGITLRCCPSLFMLKDIMDLRKCTH"
     ]
    }
   ],
   "source": [
    "dictionary ={ 'UUU': 'F', 'CUU': 'L', 'AUU': 'I', 'GUU': 'V', \\\n",
    "                    'UUC': 'F', 'CUC': 'L', 'AUC': 'I', 'GUC': 'V', \\\n",
    "                    'UUA': 'L', 'CUA': 'L', 'AUA': 'I', 'GUA': 'V', \\\n",
    "                    'UUG': 'L', 'CUG': 'L', 'AUG': 'M', 'GUG': 'V', \\\n",
    "                    'UCU': 'S', 'CCU': 'P', 'ACU': 'T', 'GCU': 'A', \\\n",
    "                    'UCC': 'S', 'CCC': 'P', 'ACC': 'T', 'GCC': 'A', \\\n",
    "                    'UCA': 'S', 'CCA': 'P', 'ACA': 'T', 'GCA': 'A', \\\n",
    "                    'UCG': 'S', 'CCG': 'P', 'ACG': 'T', 'GCG': 'A', \\\n",
    "                    'UAU': 'Y', 'CAU': 'H', 'AAU': 'N', 'GAU': 'D', \\\n",
    "                    'UAC': 'Y', 'CAC': 'H', 'AAC': 'N', 'GAC': 'D', \\\n",
    "                    'UAA': 'Stop', 'CAA': 'Q', 'AAA': 'K', 'GAA': 'E', \\\n",
    "                    'UAG': 'Stop', 'CAG': 'Q', 'AAG': 'K', 'GAG': 'E', \\\n",
    "                    'UGU': 'C', 'CGU': 'R', 'AGU': 'S', 'GGU': 'G', \\\n",
    "                    'UGC': 'C', 'CGC': 'R', 'AGC': 'S', 'GGC': 'G', \\\n",
    "                    'UGA': 'Stop', 'CGA': 'R', 'AGA': 'R', 'GGA': 'G', \\\n",
    "                    'UGG': 'W', 'CGG': 'R', 'AGG': 'R', 'GGG': 'G'}\n",
    "\n",
    "\n",
    "file = open(\"rosalind_splc.txt\", \"r\")\n",
    "\n",
    "def seqinput(file):\n",
    "    sequences = list()\n",
    "    result = \"\"       \n",
    "    for line in file:\n",
    "        if \">\" in line:\n",
    "            sequences.append(result)\n",
    "            result = \"\"\n",
    "        \n",
    "        elif \"\\n\" not in line:\n",
    "                result += line\n",
    "        else:\n",
    "            result += line[:len(line) - 1]\n",
    "    \n",
    "    sequences.append(result)\n",
    "    sequences.remove('')\n",
    "    return sequences\n",
    "\n",
    "sequences = seqinput(file)\n",
    "dna_string = sequences[0]\n",
    "introns = sequences[1:]\n",
    "introns.sort(reverse = True)\n",
    "\n",
    "for intron in introns:\n",
    "    dna_string = dna_string.replace(intron,\"\")\n",
    "dna_string = dna_string.replace(\"T\",\"U\")\n",
    "\n",
    "for i in range(0, len(dna_string) - 3, 3):\n",
    "    print(dictionary[dna_string[i: i +3]],end='')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- pmch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7196272704742491028060152791040000000\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO                      \n",
    "from math import factorial                 \n",
    "sequence = ''                              \n",
    "myfile = open('rosalind_pmch.txt', 'r')     \n",
    "for record in SeqIO.parse(myfile, 'fasta'):\n",
    "    sequence = str(record.seq)             \n",
    "a=sequence.count('A')\n",
    "c=sequence.count('C')\n",
    "matchings = factorial(a) * factorial(c)  \n",
    "    \n",
    "print(matchings)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- pper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "934400\n"
     ]
    }
   ],
   "source": [
    "from math import factorial\n",
    "\n",
    "n=96\n",
    "k=9\n",
    "perm = factorial(n)/(factorial(k)*factorial(n-k))% 1000000 * factorial(k)% 1000000\n",
    "pert= (perm) \n",
    "\n",
    "permm=int(pert)\n",
    "print(permm)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37\n"
     ]
    }
   ],
   "source": [
    "f = open('rosalind_tree.txt', 'r')\n",
    "str1 = f.readlines()\n",
    "adjacentlist = []\n",
    "for line in str1:\n",
    "    myline=line.split()\n",
    "    adjacentlist.append(myline)\n",
    "mylen= len(adjacentlist)\n",
    "print(int(adjacentlist[0][0]) - mylen)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATTAGACCTGCCGGAATAC\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "def overlap(a,b,min_lenght=3):\n",
    "    start = 0\n",
    "    while True:\n",
    "        start= a.find(b[:min_lenght],start)\n",
    "        if start == -1:\n",
    "            return 0\n",
    "        if b.startswith(a[start:]):\n",
    "            return len(a)- start\n",
    "        start += 1\n",
    "\n",
    "\n",
    "from Bio import SeqIO\n",
    "import itertools\n",
    "\n",
    "def scs(ss):\n",
    "    shortestsup= None\n",
    "    for each in itertools.permutations(ss):\n",
    "        supshort=each[0]\n",
    "        for i in range(len(ss)-1):\n",
    "            \n",
    "             overlaplenght= overlap(each[i],each[i+1],min_lenght=1) \n",
    "             supshort += each[i+1][overlaplenght:]\n",
    "        if shortestsup is None or len(supshort)< len(shortestsup):\n",
    "            shortestsup= supshort\n",
    "    return(shortestsup)    \n",
    "print(scs(['ATTAGACCTG','CCTGCCGGAA','AGACCTGCCG','GCCGGAATAC']))\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- trans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.591549295774648\n"
     ]
    }
   ],
   "source": [
    "def tran(path):\n",
    "    with open(path,'r') as fp:\n",
    "        sequences=[]\n",
    "        result=''\n",
    "        for f in fp:\n",
    "            if f[0]=='>':\n",
    "                sequences.append(result)\n",
    "                result=''\n",
    "            else:result+=f.strip()\n",
    "    return count_transition_transversion(sequences[1],result)\n",
    "\n",
    "def TransTransvRatio(s1,s2): \n",
    "    transCount=0\n",
    "    transvCount=0\n",
    "    mismatch_pairs=[ ]\n",
    "    for i in range(len(s1)):\n",
    "     pairs=0\n",
    "    if s1[i] == s2[i]:\n",
    "        pairs +=1\n",
    "    if s1[i] != s2[i]:  \n",
    "        pairs=s1[i]+s2[i]\n",
    "        mismatch_pairs.append(pairs)\n",
    "    for pair in mismatch_pairs:\n",
    "\n",
    "        if pair ==\"AG\" or pair==\"GA\" or pair==\"CT\" or pair==\"TC\":\n",
    "         transCount+=1\n",
    "        if pair == \"AT\" or pair == \"TA\" or pair == \"CG\" or pair == \"GC\":\n",
    "         transvCount+=1\n",
    "    return(transCount/transvCount)\n",
    "\n",
    "def count_transition_transversion(s,t):\n",
    "    transversion={'A':['T','C'],'G':['T','C'],'T':['A','G'],'C':['A','G']}\n",
    "    transition={'A':'G','G':'A','T':'C','C':'T'}\n",
    "    n=len(s)\n",
    "    return sum([1 if transition[s[i]]==t[i] else 0 for i in range(n)])/sum([1 if t[i] in transversion[s[i]] else 0 for i in range(n)])\n",
    "print(tran(r'rosalind_tran.txt'))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- SSEQ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3 4 13 15 21 30 32 41 46 49 57 58 60 71 72 75 76 77 80 82 84 97 98 100 115 123 126 127 129 139 144 145 154 162 164 167 168 169 184 193 200 202 203 204 212\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "file =open('rosalind_sseq.txt', 'r')\n",
    "records = SeqIO.parse(file, 'fasta')\n",
    "list_seq = [record.seq for record in records]\n",
    "S_string1 = list_seq[0]\n",
    "T_string2 = list_seq[1]\n",
    "\n",
    "lists =list()\n",
    "list_a = list()\n",
    "for i in range(len(T_string2)):\n",
    "    positions = list()\n",
    "    for j in range(len(S_string1)):\n",
    "        if T_string2[i] != S_string1[j]:\n",
    "            continue\n",
    "        if T_string2[i] == S_string1[j]:\n",
    "            positions.append(j+1)\n",
    "    lists.append(positions)\n",
    "    list_a += positions\n",
    "\n",
    "\n",
    "sortedlist = sorted(set(list_a))\n",
    "mylen=len(sortedlist)\n",
    "prints = []\n",
    "j = 0\n",
    "for i in range(mylen):\n",
    "        if j < len(lists) and sortedlist[i] in lists[j]:\n",
    "            prints.append(sortedlist[i])\n",
    "            j += 1\n",
    "print(' '.join(str(i) for i in prints))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- mprt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P02974\n",
      "67 68 121\n",
      "P23185\n",
      "71 77\n",
      "P19823\n",
      "118 445\n",
      "P00749\n",
      "322\n",
      "P01217\n",
      "80 106\n",
      "P20268\n",
      "26 126 268\n",
      "A1TJ10\n",
      "86 251\n",
      "O14977\n",
      "9\n",
      "P80370\n",
      "100\n",
      "P05155\n",
      "25 69 81 238 253 352\n",
      "P07359\n",
      "37 175 362 398\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import requests\n",
    "url = 'https://www.uniprot.org/uniprot/'\n",
    " \n",
    "\n",
    "with open(\"rosalind_mprt.txt\") as file:\n",
    " seqIDs = file.read().replace(\"\\n\", \" \").split()\n",
    " sequences = {}\n",
    "\n",
    "for proID in seqIDs:\n",
    "    goToURL = url+proID+\".fasta\"\n",
    "    response = requests.get(goToURL)\n",
    "    sequences[proID] = (response.text.split(\"\\n\"))\n",
    "    sequences[proID] = \"\".join(sequences[proID][1::])\n",
    "\n",
    "\n",
    "def N_gly_motif(ID, sequence):\n",
    "    sequence = list(sequence)\n",
    "    global result\n",
    "    result = []\n",
    "    for i in range(0, len(sequence)-3):\n",
    "        seq = sequence[i:i+4]\n",
    "        if (seq[0] == \"N\") and (seq[2] == \"S\" or seq[2] == \"T\") and (seq[1] and seq[3] != \"P\"):\n",
    "            result.append(i+1)\n",
    "\n",
    "\n",
    "for key, value in sequences.items():\n",
    "    N_gly_motif(key, value)\n",
    "    if not result:\n",
    "        continue\n",
    "    else:\n",
    "        print(key)\n",
    "        print(*result)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.1 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.1"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "afb734500600fd355917ca529030176ea0ca205570884b88f2f6f7d791fd3fbe"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
